channeldb+lnwallet: store updates the peer should sign under new key

Crypt-iQ opened the pull request:
This PR fixes a long-standing force close bug. Currently, when we send a revoke for our current commitment, we store all of the remote log updates that we need to sign for. Similarly, this PR stores the local log updates that the remote needs to sign for once we receive a revocation.


Fixes #4263 
Fixes #4435 
Fixes #4134
Fixes #4137
Fixes #4134 
Fixes #4246

Roasbeef commented on channeldb/channel.go:
```diff
@@ -72,6 +72,11 @@ var (
 	// for in one of our remote commits.
 	unsignedAckedUpdatesKey = []byte("unsigned-acked-updates-key")
 
+	// peerUnsignedAckedUpdatesKey is an entry in the channel bucket that
```
To avoid ambiguity with the above key, how about: `remoteUnsignedLocalUpdates`?

Roasbeef commented on channeldb/channel.go:
```diff
@@ -1440,6 +1445,38 @@ func (c *OpenChannel) UpdateCommitment(newCommitment *ChannelCommitment,
+		}
+
+		r := bytes.NewReader(updateBytes)
+		updates, err := deserializeLogUpdates(r)
```
Error ignored.

Roasbeef commented on channeldb/channel.go:
```diff
@@ -1440,6 +1445,38 @@ func (c *OpenChannel) UpdateCommitment(newCommitment *ChannelCommitment,
+			lIdx := upd.LogIndex
+
+			// Check if it's not on our local commitment
+			if lIdx >= newCommitment.LocalLogIndex {
```
No need for the intermediate variable here? (assuming if fits within the char limit)

Roasbeef commented on channeldb/channel.go:
```diff
@@ -2065,6 +2102,38 @@ func (c *OpenChannel) UnsignedAckedUpdates() ([]LogUpdate, error) {
 
+// PeerUnsignedAckedUpdates retrieves the persisted, unsigned acked local log
+// updates that the peer still needs to sign for.
+func (c *OpenChannel) PeerUnsignedAckedUpdates() ([]LogUpdate, error) {
```
Comment about key name above applies here too.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -1904,6 +2014,43 @@ func (lc *LightningChannel) restorePendingRemoteUpdates(
+			return err
+		}
+
+		if payDesc.EntryType == Add {
```
The comment in `localLogUpdateToPayDesc` claims an add will never be returned.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -1904,6 +2014,43 @@ func (lc *LightningChannel) restorePendingRemoteUpdates(
+			lc.localUpdateLog.restoreUpdate(payDesc)
+
+		default:
+			// settle/fail/malformed fail
```
Comment should be a complete sentence.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -1904,6 +2014,43 @@ func (lc *LightningChannel) restorePendingRemoteUpdates(
+
+		default:
+			// settle/fail/malformed fail
+			lc.localUpdateLog.restoreUpdate(payDesc)
```
Alternative way to express this logic: 
```go
lc.localUpdateLog.restoreUpdate(payDesc)

if payDesc.EntryType != FeeUpdate {
    lc.remoteUpdateLog.markHtlcModified(payDesc.ParentIndex)
}
```

Crypt-iQ replied:
Done

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -4573,6 +4720,60 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 		}
 	}
 
+	var localPeerUpdates []channeldb.LogUpdate
```
Should move this into its own function as this method is already pretty long.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -4573,6 +4720,60 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
+				LogIndex: pd.LogIndex,
+			}
+
+			// don't save adds
```
Comment should be a complete sentence.

Crypt-iQ replied:
Mb, fixed.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -4573,6 +4720,60 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
+		pd := e.Value.(*PaymentDescriptor)
+
+		if pd.LogIndex < remoteMessageIndex && pd.LogIndex >= localMessageIndex {
+			// This is a settle/fail that is on the remote commitment, but
```
Style nit: move the comment to reside _above_ the if statement?

Crypt-iQ replied:
Should this always be the case? What if there's an `else`?

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -4573,6 +4720,60 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
+			}
+
+			// don't save adds
+			if pd.EntryType == Add {
```
This can be moved to reside outside the if statement so we can skip them early.

Crypt-iQ replied:
done

Roasbeef commented:
New sporadic itests log error failure (looks to be unrelated to this PR): 
> <time> [ERR] HSWC: ChannelLink(<chan>): outgoing htlc(<hex>) has insufficient fee: expected 612500, got 1112

Roasbeef commented on channeldb/channel.go:
```diff
@@ -79,6 +79,11 @@ var (
+	// remoteUnsignedLocalUpdatesKey is an entry in the channel bucket that
+	// contains the local updates that the remote party has acked, but
+	// has not yet signed for in onr of their local commits.
+	remoteUnsignedLocalUpdatesKey = []byte("remote-unsigned-local-updates-key")
```
👍

Roasbeef commented on channeldb/channel.go:
```diff
@@ -2073,6 +2112,38 @@ func (c *OpenChannel) UnsignedAckedUpdates() ([]LogUpdate, error) {
+			tx, c.IdentityPub, &c.FundingOutpoint, c.ChainHash,
+		)
+		switch err {
+		case nil:
```
Could make this slightly more explicit by adding a `break` statement below this case. Otherwise, at a glance, it looks like it can fall through to the lower case statement, causing it to return `nil` when the bucket is actually found.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -1904,6 +2014,34 @@ func (lc *LightningChannel) restorePendingRemoteUpdates(
+
+		lc.localUpdateLog.restoreUpdate(payDesc)
+
+		if payDesc.EntryType != FeeUpdate {
```
Could add a comment here that only adds are stored under this key, so we only need this special case for fee updates.

Crypt-iQ replied:
PTAL at comment

Crypt-iQ commented on channeldb/db_test.go:
```diff
@@ -398,7 +398,7 @@ func TestRestoreChannelShells(t *testing.T) {
 		t.Fatalf("able to mutate restored channel")
 	}
-	err = channel.AdvanceCommitChainTail(nil)
+	err = channel.AdvanceCommitChainTail(nil, nil)
```
Yes we should

Crypt-iQ replied:
Not entirely sure how this would look.

Roasbeef commented:
The failing race condition test look unrelated: 
```
--- FAIL: TestFundingManagerMaxPendingChannels (5.93s)
    fundingmanager_test.go:2845: peer did not send FundingLocked message
```

Crypt-iQ commented on lnwallet/channel.go:
```diff
@@ -1527,6 +1527,100 @@ func (lc *LightningChannel) logUpdateToPayDesc(logUpdate *channeldb.LogUpdate,
+	remoteUpdateLog *updateLog, commitHeight uint64) (*PaymentDescriptor,
+	error) {
+
+	// The logUpdate here won't ever be an Add.
```
PTAL at comment

Crypt-iQ commented on lnwallet/channel.go:
```diff
@@ -1527,6 +1527,100 @@ func (lc *LightningChannel) logUpdateToPayDesc(logUpdate *channeldb.LogUpdate,
+		}, nil
+
+	default:
+		return nil, errors.New("unknown message type")
```
Done

Crypt-iQ commented:
> The failing race condition test look unrelated:
> 
> ```
> --- FAIL: TestFundingManagerMaxPendingChannels (5.93s)
>     fundingmanager_test.go:2845: peer did not send FundingLocked message
> ```

Yes race condition is unrelated. Addressed all comments but Conner's about including tests for `AdvanceCommitChainTail`.

halseth commented on channeldb/channel.go:
```diff
@@ -2102,7 +2173,9 @@ func (c *OpenChannel) InsertNextRevocation(revKey *btcec.PublicKey) error {
 // remote party to the revocation log, and promote the current pending
 // commitment to the current remote commitment.
-func (c *OpenChannel) AdvanceCommitChainTail(fwdPkg *FwdPkg) error {
+func (c *OpenChannel) AdvanceCommitChainTail(fwdPkg *FwdPkg,
```
nit: godoc should be updated to explain the new argument

Crypt-iQ replied:
PTAL @ comment

halseth commented on lnwallet/channel.go:
```diff
@@ -1527,6 +1527,101 @@ func (lc *LightningChannel) logUpdateToPayDesc(logUpdate *channeldb.LogUpdate,
+	case *lnwire.UpdateFulfillHTLC:
+		ogHTLC := remoteUpdateLog.lookupHtlc(wireMsg.ID)
+
+		if ogHTLC.addCommitHeightRemote == 0 {
```
is this necessary? Didn't we fix that in the previous PR?

Crypt-iQ replied:
This is the opposite side, I moved it to where the other side lives in `restoreStateLogs`

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -1873,7 +1975,12 @@ func (lc *LightningChannel) restoreStateLogs(
-	return nil
+	// Restore unsigned acked local log updates so we expect the peer to
+	// sign for them.
+	err = lc.restorePeerLocalUpdates(
```
Style nit: can just return the method directly here since it only has a single return value.
